research/zkemail/README.md

```diff
@@ -0,0 +1,273 @@
+
+
+## Enigmatic 255-decimal
+
+The cornerstone part of the ZK Email system is the ZK regex compiler. It empowers a ZK circuit to support the processing of input signals with beloved and cherished regular expressions. The ZK regexes allow not only to constrain the input to satisfy the regex but additionally to produce a revealing array, exposing the part of the input matching specific sub-regexes.
+
+Let’s look at the concrete example of a regex fed to the ZK regex compiler. This one comprises several sub-regexes and should match the whole `dkim-signature` header in an email but reveal only the timestamp value for the `;t=ts_value` tag in the header.
+
+```
+{
+  "parts": [
+    {
+      "is_public": false,
+      "regex_def": "(\r\n|^)dkim-signature:"
+    },
+    {
+      "is_public": false,
+      "regex_def": "([a-z]+=[^;]+; )+t="
+    },
+    {
+      "is_public": true,
+      "regex_def": "[0-9]+"
+    },
+    {
+      "is_public": false,
+      "regex_def": ";"
+    }
+  ]
+}
+```
+
+The ZK Email relies on regular expressions prefixed with `(\r\n|^)header_name:` for extracting various headers from an email, such as the `from`, `subject`, `dkim-signature` headers.
+
+When we compile a regular expression with the compiler, it generates a Circom circuit that embodies a Deterministic Finite Automaton (DFA) satisfying the regular expression. Any regular expression can be converted into an equivalent DFA. An input string matches the regular expression if it transitions the DFA to the accept state.
+
+Let’s have the `^a` regular expression as a toy example. When generating a circuit for the regular expression, the compiler makes a DFA with three states, with state `2` being the accepted state.
+
+What’s more essential for understanding the vulnerability identified is that the compiler injects `255` decimal as the first value to the `in` array before the user’s input. This one is called an “invalid” decimal, as conceived by the ZK regex compiler’s code, and shouldn’t interfere with the user’s input, indicating the beginning of the string.
+
+```
+var num_bytes = msg_bytes+1;
+signal in[num_bytes];
+in[0]<==255;
+for (var i = 0; i < msg_bytes; i++) {
+	in[i+1] <== msg[i];
+}
+```
+
+Further, in the main loop, the circuit handles the transition between states of the DFA based on the input `in[i]` and the current state. Initially, the DFA is in state `0` and transitions to state `1` when it receives the 255-decimal input, which denotes the beginning of the string or `^`.
+
+```
```

Comment:
Could you add markers to the code lines (in the google doc they're bold)? We can use a line comment here, `// *`.